src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncMethodBuilder.cs

```diff
@@ -894,7 +895,7 @@ internal static Task<TResult> GetTaskForResult(TResult result)
                 // For Boolean, we cache all possible values.
                 if (typeof(TResult) == typeof(bool)) // only the relevant branches are kept for each value-type generic instantiation
                 {
-                    bool value = (bool)(object)result;
+                    bool value = (bool)(object?)result;
```

Comment:
Just to learn, why are this double casts needed? why not directly into `bool`?

Reply:
The compiler won't let you cast directly from `T` to `bool` as it doesn't know what `T` is and whether such a cast actually exists.  By first going to `object`, we leave it up to the runtime, effectively boxing the `T` (which the compiler allows) and then unboxing it (which the compiler allows, leaving verification up to the runtime).  The JIT then has an optimization to remove the boxing for this specific pattern.